### DISCUSSION TITLE
.optional(isOptional = true)

### DISCUSSION THREAD

**Discussion by binajmen** [colinhacks/zod#1116] [Ideas]
Hi,

I'd like to conditionally set some fields optional.

Right now, it does not seem we can specify a boolean value to `.optional()`, although that would be very handy:

```ts
function myFunction(needLastName: boolean) {
  const user = z.object({
    firstName: z.string(),
    lastName: z.string().optional(needLastName),
  });
  ...
}
```

If `needLastName = true`: the field is indeed optional.
If `needLastName = false`: the `.optional()` is "dismissed" (field is in fact required).

Such a change should be non-breaking as a default value could be set to true.

Or perhaps there's another way to achieve the same result?

**Comment by scotttrinh** [comment]
You can achieve this with a ternary, but beware the inferred type signature since the type changes at runtime. It's likely that what you want here is two different schemas or some kind of union:

```ts
    lastName: needLastName ? z.string() : z.string().optional(),
```
